cozy/cost_model.py

```diff
@@ -209,6 +280,80 @@ def worst_case_cardinality(e : Exp) -> DominantTerm:
         return DominantTerm.ONE
     return DominantTerm.N
 
+def _maintenance_cost(e : Exp, solver : ModelCachingSolver, op : Op, freebies : [Exp] = []):
+    """Determines the maintenance cost of an expression with regards to its
+    freebies and ops.
+
+    The cost is the result of mutating the expression and getting the storage
+    size of the difference between the mutated expression and the original.
+    """
+    e_prime = mutate(e, op.body)
+    if solver.valid(EEq(e, e_prime)):
+        return ZERO
+
+    h = extension_handler(type(e.type))
+    if h is not None:
+        return h.maintenance_cost(e,
+                solver=solver,
+                op=op,
+                freebies=freebies,
+                storage_size=storage_size,
+                maintenance_cost=_maintenance_cost)
+
+    if is_scalar(e.type):
+        return storage_size(e, freebies)
+    elif isinstance(e.type, TBag) or isinstance(e.type, TSet):
+        things_added = storage_size(
+                EBinOp(e_prime, "-", e).with_type(e.type), freebies).with_type(INT)
+        things_remov = storage_size(
+                EBinOp(e, "-", e_prime).with_type(e.type), freebies).with_type(INT)
+
+        return ESum([things_added, things_remov])
+    elif isinstance(e.type, TList):
+            return storage_size(e_prime, freebies)
+    elif isinstance(e.type, TMap):
+        keys = EMapKeys(e).with_type(TBag(e.type.k))
+        vals = EMap(
+            keys,
+            mk_lambda(
+                e.type.k,
+                lambda k: EMapGet(e, k).with_type(e.type.v))).with_type(TBag(e.type.v))
+
+        keys_prime = EMapKeys(e_prime).with_type(TBag(e_prime.type.k))
+        vals_prime = EMap(
+            keys_prime,
+            mk_lambda(
+                e_prime.type.k,
+                lambda k: EMapGet(e_prime, k).with_type(e_prime.type.v))).with_type(TBag(e_prime.type.v))
+
+        keys_added = storage_size(
+            EBinOp(keys_prime, "-", keys).with_type(keys.type), freebies).with_type(INT)
+        keys_rmved = storage_size(
+            EBinOp(keys, "-", keys_prime).with_type(keys.type), freebies).with_type(INT)
+
+        vals_added = storage_size(
+            EBinOp(vals_prime, "-", vals).with_type(vals.type), freebies).with_type(INT)
+        vals_rmved = storage_size(
+            EBinOp(vals, "-", vals_prime).with_type(vals.type), freebies).with_type(INT)
+
+        keys_difference = ESum([keys_added, keys_rmved])
+        vals_difference = ESum([vals_added, vals_rmved])
+        return EBinOp(keys_difference, "*", vals_difference).with_type(INT)
+
+    else:
+        raise NotImplementedError(repr(e.type))
+
+def maintenance_cost(e : Exp, solver : ModelCachingSolver, op : Op, freebies : [Exp] = []):
+    """This method calulates the result over all expressions that are EStateVar """
+    res = ZERO
+    for x in all_exps(e):
+        if isinstance(x, EStateVar):
+            res = ESum([
```

Comment:
Calling `ESum` like this with only two arguments has little effect. You can simplify this code and get more benefit from `ESum` using a comprehension:

```python
return ESum([
    _maintenance_cost(e=x.e, solver=solver, op=op, freebies=freebies)
    for x in all_exps(e)
    if isinstance(x, EStateVar)])
```

(In this case, since `ESum` sees _all_ the expressions being added, it can build a more balanced syntax tree. Also, it handles the zero case for you.)